DCOS-59167 [DS] [Spark Operator] Make metrics reporting configurable and add ServiceMonitor/scraping for the Operator at the deployment step

VladimirKravtsov opened the pull request:
### What changes were proposed in this pull request?
Resolves [DCOS-59167 [DS] [Spark Operator] Make metrics reporting configurable and add ServiceMonitor/scraping for the Operator at the deployment step](https://jira.mesosphere.com/browse/DCOS-59167).

### Why are the changes needed?
To avoid `kudo-spark-operator` users configuration burden in order to get theirs Spark Applications telemetries metrics scraped by Prometheus.

### How were the changes tested?
In order to verify the changes, please, try to follow steps, described in MONITORING.md file, which was added within the PR.


**Important! Due to the [bug](https://github.com/kudobuilder/kudo/issues/913) in KUDO 0.7.3, `ServiceMonitor` is *not* became part of the operator. It will be fixed once aforementioned issue will be resolved.**

Also, because [the bug](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/pull/649/files) was resolved in `spark-on-k8s-operator`, I reverted back default app metrics port on 8090.

alembiewski commented on MONITORING.md:
````diff
@@ -0,0 +1,28 @@
+           port: 8090
+     ```  
+   - if you would like use other than 8090 port for metrics exporting, you must pass appropriate parameter during `kudo-spark-operator` installation `-p appMetricsPort=<desired_port>` 
+   - Mark your application by the label 
````
```suggestion
   - Mark your application with the label 
```

akirillov replied:
```suggestion
   - specify additional labels for the application to make it discoverable by `ServiceMonitor`
```

alembiewski commented on MONITORING.md:
```diff
@@ -0,0 +1,28 @@
+The `kudo-spark-operator` is able to seamlessly integrated with Prometheus, which is installed within [Prometheus operator](https://github.com/coreos/prometheus-operator).
```
```suggestion
The `kudo-spark-operator` integrates seamlessly with Prometheus, which is installed within [Prometheus operator](https://github.com/coreos/prometheus-operator).
```

akirillov replied:
```suggestion
By default, KUDO Spark Operator supports metrics reporting and integrates with [Prometheus operator](https://github.com/coreos/prometheus-operator).
```

akirillov commented on MONITORING.md:
```diff
@@ -0,0 +1,28 @@
+The `kudo-spark-operator` is able to seamlessly integrated with Prometheus, which is installed within [Prometheus operator](https://github.com/coreos/prometheus-operator).
+Integration with other Prometheus distributions, like kube-prometheus, wasn't tested. 
```
```suggestion
```

akirillov commented on MONITORING.md:
```diff
@@ -0,0 +1,28 @@
+The `kudo-spark-operator` is able to seamlessly integrated with Prometheus, which is installed within [Prometheus operator](https://github.com/coreos/prometheus-operator).
+Integration with other Prometheus distributions, like kube-prometheus, wasn't tested. 
+
+The `prometheus-operator` uses crafted services discovery approach, introducing `ServiceMonitor` kind. 
```
Prometheus Operator relies on `ServiceMonitor` kind which describes the set of targets to be monitored. KUDO Spark Operator configures `ServiceMonitor`s for both the Operator and submitted Spark Applications automatically when monitoring is enabled.

VladimirKravtsov replied:
Thanks!

akirillov replied:
didn't work as a suggestion for some reason, do you think we still need the sentence above?

VladimirKravtsov replied:
it removed

akirillov commented on MONITORING.md:
```diff
@@ -0,0 +1,28 @@
+The `prometheus-operator` uses crafted services discovery approach, introducing `ServiceMonitor` kind. 
+But the `kudo-spark-operator` take its configuration burden on itself.
+
+#### How to get `kudo-spark-operator` and app's metrics available for scrape by Prometheus:
```
```suggestion
#### Configuring Spark Operator and Spark Application metrics export to Prometheus
```

akirillov commented on MONITORING.md:
```diff
@@ -0,0 +1,28 @@
+
+#### How to get `kudo-spark-operator` and app's metrics available for scrape by Prometheus:
+1) Ensure `prometheus-operator` is installed on your Kubernetes cluster.
+1) Install the `kudo-spark-operator` with `-p enableMetrics=true` parameter. (But it's true by default, so typically no need to specify it explicitly)
```
```suggestion
1) Install the KUDO Spark Operator. Metrics reporting is enabled by default and can be disabled by modifying `enableMetrics` parameter.
```

VladimirKravtsov replied:
Thanks, Anton!

akirillov commented on MONITORING.md:
```diff
@@ -0,0 +1,28 @@
+1) Install the `kudo-spark-operator` with `-p enableMetrics=true` parameter. (But it's true by default, so typically no need to specify it explicitly)
+1) Composing your Spark Application yaml:
+   - use Spark image with JMXPrometheus exporter jar on the board i.e. `gcr.io/spark-operator/spark:v2.4.4-gcs-prometheus` 
+   - enable driver/executors monitoring by adding the yaml piece into `spec` section:
```
```suggestion
   - enable Driver and Executors metrics reporting by adding the following configuration into `SparkApplication` `spec` section:
```

akirillov commented on MONITORING.md:
````diff
@@ -0,0 +1,28 @@
+           jmxExporterJar: "/prometheus/jmx_prometheus_javaagent-0.11.0.jar"
+           port: 8090
+     ```  
+   - if you would like use other than 8090 port for metrics exporting, you must pass appropriate parameter during `kudo-spark-operator` installation `-p appMetricsPort=<desired_port>` 
````
```suggestion
   - if you would like to use a different port for metrics exporting, it should be provided via `-p appMetricsPort=<desired_port>` during KUDO Spark Operator installation
```

akirillov commented on MONITORING.md:
````diff
@@ -0,0 +1,28 @@
+         exposedMetrics: true
+     ```
+
+Full blown example you can find in [spark-application.yaml](specs/spark-application.yaml).
````
```suggestion
Full configuration example is available in [specs/spark-application.yaml](specs/spark-application.yaml).
```

VladimirKravtsov replied:
Could you, pls, shine light on why the wording better?

akirillov replied:
Sure, here are some points:
* original sentence lacks the subject (example of what?)
* word ordering issue (should be 'you can find the example')
* we avoid using pronouns ('you') in the service docs
* 'Full blown' doesn't suit the context here (see [Merriam-Webster](https://www.merriam-webster.com/dictionary/full-blown) definition and synonyms)
* `spark-application.yaml` -> `specs/spark-application.yaml` in the link name provides the location of the file in the project,  not just the link

Does it make sense?

VladimirKravtsov replied:
@akirillov, it's sweet :-) Thanks for the explanation. It makes sense. Just a couple of comments from my side:  
> original sentence lacks the subject (example of what?)

neat!

> we avoid using pronouns ('you') in the service docs

Good point. And If you have somewhere guide/list of best practices like that, which we should follow. I'd appreciate if you will share it with entire team.

> 'Full blown' doesn't suit the context here (see Merriam-Webster definition and synonyms)

I do not insist to use `Full blown' phrase here :-) just to debate - as per [Collins dictionary](https://www.collinsdictionary.com/dictionary/english/full-blown) it also has meaning "fully grown or developed;", which is suitable for this context, imo.

> spark-application.yaml -> specs/spark-application.yaml in the link name provides the location of the file in the project, not just the link

mb it makes sense for better transparency. Beside of that user always can see location in url or, after click, in github ui.

akirillov replied:
you can check other services' docs created by the team for the inspiration regarding the style and format e.g. [Data Science Engine](https://docs.d2iq.com/mesosphere/dcos/services/data-science-engine/1.0.0/) or [DCOS Spark](https://docs.d2iq.com/mesosphere/dcos/services/spark/2.9.0-2.4.3/). I think Docs team has some guidelines but usually, we follow review suggestions.

RE: 'full blown', I don't think it's wrong it's just doesn't suit the context: 'fully grown configuration example' or 'fully developed configuration example' (developed not in the software sense). I'd expect 'full' or 'complete' here, but it's up to you.

VladimirKravtsov replied:
ok, thanks, make sense

akirillov commented on README.md:
```diff
@@ -75,6 +75,8 @@ make install NAMESPACE=test-namespace
 
 ### Submitting Spark Application
 
+To get started with your app monitoring, please, see also [MONITORING.md](MONITORING.md) 
```
it makes sense to place this reference at the end of the section

VladimirKravtsov replied:
done

akirillov commented on kudo-operator/operator.yaml:
```diff
@@ -12,14 +12,17 @@ url: https://spark.apache.org/
   deploy:
     resources:
-      - crds.yaml
+      - spark-operator-crd.yaml
```
shouldn't it be plural: `spark-operator-crds.yaml` because there're multiple CRDs defined in it?

akirillov commented on kudo-operator/params.yaml:
```diff
@@ -47,9 +47,14 @@ enableMetrics:
   displayName: "Enable metrics"
 
 metricsPort:
```
the naming is not consistent with `appMetricsPort`. What if we rename it to `operatorMetricsPort`?

VladimirKravtsov replied:
done

akirillov commented on scripts/install_operator.sh:
```diff
@@ -29,7 +29,7 @@ EOF
-# Create ServiceMonitor (see prometheus-operator docs) for Spark
+
+# Create ServiceMonitor (see prometheus-operator docs) for Spark.
+# TODO: make the service-monitor.yaml part of kudo-spark-operator once the issue https://github.com/kudobuilder/kudo/issues/913 will be fixed.
```
can we please make this information available in `monitoring.md` until this bug is not fixed?

VladimirKravtsov replied:
Ok, make sense

akirillov replied:
has it been added to `monitoring.md`? can't find it there

VladimirKravtsov replied:
https://github.com/mesosphere/kudo-spark-operator/pull/17/files#diff-fa49808299ade494be7376593e58101fR12-R30

VladimirKravtsov commented on MONITORING.md:
```diff
@@ -0,0 +1,28 @@
+Integration with other Prometheus distributions, like kube-prometheus, wasn't tested. 
+
+The `prometheus-operator` uses crafted services discovery approach, introducing `ServiceMonitor` kind. 
+But the `kudo-spark-operator` take its configuration burden on itself.
```
Thanks, Sasha

VladimirKravtsov commented on MONITORING.md:
```diff
@@ -0,0 +1,28 @@
+
+#### How to get `kudo-spark-operator` and app's metrics available for scrape by Prometheus:
+1) Ensure `prometheus-operator` is installed on your Kubernetes cluster.
+1) Install the `kudo-spark-operator` with `-p enableMetrics=true` parameter. (But it's true by default, so typically no need to specify it explicitly)
```
Fixed by Anton's suggestion apply.

VladimirKravtsov commented on MONITORING.md:
```diff
@@ -0,0 +1,28 @@
+1) Ensure `prometheus-operator` is installed on your Kubernetes cluster.
+1) Install the `kudo-spark-operator` with `-p enableMetrics=true` parameter. (But it's true by default, so typically no need to specify it explicitly)
+1) Composing your Spark Application yaml:
+   - use Spark image with JMXPrometheus exporter jar on the board i.e. `gcr.io/spark-operator/spark:v2.4.4-gcs-prometheus` 
```
I agree, here should be `mesosphere/spark` image. The reason why I didn't put it here this time - the aforementioned image isn't verified in terms of metrics exporting. I remember you have done good job adding Prometheus configs there, but the image should be assembled, pushed into Docker Hub and checked that the metrics exported. I'm going to do in scope of other PR (introducing Operator dashboard). So, I'd like leave it as is here at this time.

alembiewski replied:
> the image should be assembled, pushed into Docker Hub and checked that the metrics exported

yes, this is basically what I've done in scope of testing #12 to verify the metrics are exported into Prometheus. The only difference is that I used my personal docker hub to do that. Since it's a readme file, I don't see any issue with specifying `mesosphere/image` here at the first place just to prevent any confusion or just to not to forget to do it later

alembiewski replied:
You can use this image `mesosphere/spark:2.4.4-bin-hadoop2.7-k8s`

VladimirKravtsov replied:
Thanks @alembiewski. As we discussed, metrics exporting is verified with this image. So it makes sense to mention it here instead of Google's one.

samvantran commented on MONITORING.md:
```diff
@@ -0,0 +1,56 @@
+Out of the box, the `kudo-spark-operator` has enabled metrics reporting. 
+By default, it supports integration with [Prometheus operator](https://github.com/coreos/prometheus-operator).
```
```suggestion
By default, it supports integration with the [Prometheus operator](https://github.com/coreos/prometheus-operator).
```

samvantran commented on MONITORING.md:
```diff
@@ -0,0 +1,56 @@
+KUDO Spark Operator configures `ServiceMonitor`s for both the Operator and submitted Spark Applications automatically 
+when monitoring is enabled.
+
+#### Configuring Spark Operator and Spark Application metrics export to Prometheus
```
```suggestion
#### Exporting Spark Operator and Spark Application metrics to Prometheus
```

samvantran commented on MONITORING.md:
````diff
@@ -0,0 +1,56 @@
+   ```yaml
+   cat <<EOF | kubectl apply -f -
+   apiVersion: monitoring.coreos.com/v1
+    kind: ServiceMonitor
````
```suggestion
   kind: ServiceMonitor
```

samvantran commented on MONITORING.md:
````diff
@@ -0,0 +1,56 @@
+           metrics-exposed: "true"
+     ```
+
+Full configuration example is available in [specs/spark-application.yaml](specs/spark-application.yaml).
````
The image in this example uses `gcr.io/spark-operator/spark:v2.4.4-gcs-prometheus`. Should we update it so that the instructions match?

samvantran replied:
Also, this example file does not have the `metrics-exposed` labels as suggested above. Might confuse users.

VladimirKravtsov replied:
> Also, this example file does not have the metrics-exposed labels as suggested above. Might confuse users.
It has :-) Take a look, pls, [the label in executor spec](https://github.com/mesosphere/kudo-spark-operator/pull/17/files#diff-94f8071a156ee28e0f01545179230631R37) and [the one in driver spec](https://github.com/mesosphere/kudo-spark-operator/pull/17/files#diff-94f8071a156ee28e0f01545179230631R26)

VladimirKravtsov replied:
> The image in this example uses gcr.io/spark-operator/spark:v2.4.4-gcs-prometheus. Should we update it so that the instructions match?

Done. https://github.com/mesosphere/kudo-spark-operator/pull/17/files#diff-94f8071a156ee28e0f01545179230631R8

samvantran replied:
So when I went through the instructions I got up to this point and then thought, "what next?" 

At this point you've showed users can install the operator and set the configs needed in the spark application to enable metrics. But the docs stop before you instruct users:

- to deploy the spark application
- find the URL for the prometheus dashboard
- or if there a `/metrics` endpoint that users can go to see which metrics are being sent?

The final conclusion of this doc should be something like 
```
Now you can view metrics by going to <url> and seeing the spark metrics being sent
```
or 
```
Now go to the prometheus dashboard at <url> and search for metrics starting with 'spark'
```
Let's be empathetic to the end users

VladimirKravtsov replied:
@samvantran very good point, thanks for catching it! 
Addressed.

samvantran commented on MONITORING.md:
````diff
@@ -0,0 +1,56 @@
+   EOF
+   ```
+1) Composing your Spark Application yaml:
+   - use Spark image with JMXPrometheus exporter jar on the board i.e. `mesosphere/spark:2.4.4-bin-hadoop2.7-k8s` 
````
```suggestion
   - use the following Spark image which includes the `JMXPrometheus` exporter jar: `mesosphere/spark:2.4.4-bin-hadoop2.7-k8s` 
```

samvantran commented on MONITORING.md:
````diff
@@ -0,0 +1,56 @@
+           port: 8090
+     ```  
+   - if it's necessary to expose metrics endpoint on other than 8090 port, then, please, 
+     1) change the port value in `SparkApplication` definition yaml (`spec.monitoring.prometheus.port`) and 
````
```suggestion
     1) change the `port` value in the `SparkApplication` yaml definition (`spec.monitoring.prometheus.port`)
```

samvantran commented on MONITORING.md:
````diff
@@ -0,0 +1,56 @@
+     ```  
+   - if it's necessary to expose metrics endpoint on other than 8090 port, then, please, 
+     1) change the port value in `SparkApplication` definition yaml (`spec.monitoring.prometheus.port`) and 
+     1) specify appropriate parameter during `kudo-spark-operator` installation - `-p appMetricsPort=<desired_port>`. 
````
Can you share a fuller cmd example? Something like:

```suggestion
     1) specify the same port when installing the `kudo-spark-operator`:  
     ```
     kubectl kudo install <operator> -p appMetricsPort=<desired_port>
     ```
```

VladimirKravtsov commented:
@alembiewski thanks for attentive review and valuable suggestions!

Apart of requested changes. I made:
* Extracted metrics Service endpoint installation to `install_operator.sh` script.
* Improved `remove_operator.py` script. (added CRDs removing as well)
* Rebased on master.

It's means that metrics exposing and scrapping are working out of the box, avoiding blockers on kudo. 
@akirillov , @samvantran PTAL

VladimirKravtsov commented:
Also, here is added utility script for remove kudo from cluster https://github.com/mesosphere/kudo-spark-operator/blob/DCOS-59167/scripts/remove_kudo.sh

samvantran commented on MONITORING.md:
````diff
@@ -0,0 +1,58 @@
+           jmxExporterJar: "/prometheus/jmx_prometheus_javaagent-0.11.0.jar"
+           port: 8090
+     ```  
+   - if it's necessary to expose metrics endpoint on other than 8090 port, then, please, 
````
```suggestion
   - if it's necessary to expose the metrics endpoint on a port other than `8090`, do the following:
```

samvantran commented on MONITORING.md:
```diff
@@ -0,0 +1,58 @@
+when monitoring is enabled.
+
+#### Exporting Spark Operator and Spark Application metrics to Prometheus
+1) Ensure `prometheus-operator` is installed on your Kubernetes cluster.
```
Can you provide a link that shows users what instructions are needed to install the prometheus-operator? And some information on how they can get to the prometheus dashboard once installed? 

Or are you expecting users to be able to access the pre-installed prom/grafana charts that come with konvoy?

VladimirKravtsov replied:
Yes, I'm expecting users install the operator to Konvoy, where all of these tools (Prometheus, Grafana) are installed.
On the other hand, we agreed to make the kudo-spark-operator able to run on any kubernetes cluster. So, it make sense provide link to an instruction how to install the prometheus-operator like [this](https://github.com/coreos/prometheus-operator#quickstart).  My only concern I don't personally tried install `prometheus-operator` on vanilla kubernetes cluster and maybe can't help a user in case of any troubles with it.

VladimirKravtsov replied:
addressed

VladimirKravtsov commented on MONITORING.md:
```diff
@@ -0,0 +1,58 @@
+
+#### Exporting Spark Operator and Spark Application metrics to Prometheus
+1) Ensure `prometheus-operator` is installed on your Kubernetes cluster.
+1) Install the KUDO Spark Operator. Metrics reporting is enabled by default and can be disabled by modifying `enableMetrics` parameter.
```
I'm expecting the user will come to the guide from README.md, where he followed instruction how to install the operator. I think we can remove this point from the instruction. I will introduce prerequisite section, where the kudo-spark-operator will be mentioned

VladimirKravtsov replied:
Addressed

VladimirKravtsov commented on MONITORING.md:
````diff
@@ -0,0 +1,58 @@
+1) Ensure `prometheus-operator` is installed on your Kubernetes cluster.
+1) Install the KUDO Spark Operator. Metrics reporting is enabled by default and can be disabled by modifying `enableMetrics` parameter.
+1) Create ServiceMonitor for Spark (see prometheus-operator docs). Take this yaml without modification - 
+   ```yaml
````
fixed

VladimirKravtsov commented:
Hey @akirillov! All your previous comments addressed now. Do you have more comments on this PR?

VladimirKravtsov commented:
@samvantran, thank you so much for careful review! All stuff is addressed, PTAL^

samvantran commented on MONITORING.md:
```diff
@@ -0,0 +1,86 @@
+         name: metrics
+     clusterIP: None
+     selector:
+       "metrics-exposed": "true"
```
missing `EOF`

samvantran commented on MONITORING.md:
````diff
@@ -0,0 +1,86 @@
+   ```
+   kubectl apply -f <path_to_the_application_yaml>   
+   ```
+   Full configuration example is available in [specs/spark-application.yaml](specs/spark-application.yaml).
````
All of these should be indented two spaces so that they are aligned properly with `- Install the SparkApplication`

samvantran commented on MONITORING.md:
````diff
@@ -0,0 +1,86 @@
+   kubectl apply -f <path_to_the_application_yaml>   
+   ```
+   Full configuration example is available in [specs/spark-application.yaml](specs/spark-application.yaml).
+1) Create the metrics endpoint service. And don't forget to modify the service port in the yaml in case you have changed it 
````
Shouldn't we set up the metrics endpoint service before submitting the spark application? Feels like this step should go right after 
```
1) Create a `ServiceMonitor` for Spark: 
```

VladimirKravtsov replied:
I placed those steps in that order just to simplify the port changing steps sequence. I tried swap them, PTAL.

samvantran replied:
👍

samvantran commented on MONITORING.md:
```diff
@@ -0,0 +1,86 @@
+
+##### Prerequisite
+* The *`prometheus-operator`*.
+If you are using Kubernetes cluster without pre-installed the `prometheus-operator` then follow 
```
```suggestion
If you don't already have the `prometheus-operator` installed on your Kubernetes cluster, you can do so by following
```

samvantran commented on MONITORING.md:
````diff
@@ -0,0 +1,86 @@
+     selector:
+       "metrics-exposed": "true"
+   ```  
+1) Now go to the prometheus dashboard at `<kubernetes_endpoint_url>/ops/portal/prometheus/graph` and search for metrics 
````
```suggestion
1) Now, go to the prometheus dashboard (e.g. `<kubernetes_endpoint_url>/ops/portal/prometheus/graph`) and search for metrics 
```

samvantran commented on MONITORING.md:
````diff
@@ -0,0 +1,86 @@
+       "metrics-exposed": "true"
+   ```  
+1) Now go to the prometheus dashboard at `<kubernetes_endpoint_url>/ops/portal/prometheus/graph` and search for metrics 
+starting with 'spark'. The Prometheus uri might be differ depends on the `prometheus-operator` installation configuration. 
````
```suggestion
starting with 'spark'. The Prometheus URI might be different depending on how you configured and installed the `prometheus-operator`. 
```
